src/main/java/com/google/webauthn/gaedemo/server/Server.java

```diff
@@ -94,4 +99,53 @@ public static Credential validateAndFindCredential(PublicKeyCredential cred, Str
 
     return credential;
   }
+
+  /**
+   * @param cred
+   * @param currentUser
+   * @param sessionId
+   * @throws ServletException
+   */
+  public static void verifyAssertion(PublicKeyCredential cred, String currentUser, String sessionId,
+                                     Credential savedCredential) throws ServletException {
+    AuthenticatorAssertionResponse assertionResponse = (AuthenticatorAssertionResponse) cred.getResponse();
+
+    Log.info("-- Verifying signature --");
+    if (!(savedCredential.getCredential().getResponse() instanceof AuthenticatorAttestationResponse)) {
+      throw new ServletException("Stored attestation missing");
+    }
+    AuthenticatorAttestationResponse storedAttData = (AuthenticatorAttestationResponse) savedCredential.getCredential()
+            .getResponse();
+
+    try {
+      PublicKey publicKey;
+      if (storedAttData.decodedObject.getAuthenticatorData().getAttData().getPublicKey() instanceof EccKey) {
+        publicKey = Crypto.getECPublicKey((EccKey) storedAttData.decodedObject.getAuthenticatorData().getAttData().getPublicKey());
+      } else {
+        publicKey = Crypto.getRSAPublicKey((RsaKey) storedAttData.decodedObject.getAuthenticatorData().getAttData().getPublicKey());
+      }
+
+      String clientDataJson = assertionResponse.getClientDataString();
+      byte[] clientDataHash = Crypto.sha256Digest(clientDataJson.getBytes());
+
+      byte[] signedBytes = Bytes.concat(storedAttData.getAttestationObject().getAuthenticatorData().getRpIdHash(),
+              new byte[] { (assertionResponse.getAuthenticatorData().isUP() == true ? (byte) 1 : (byte) 0) },
+              ByteBuffer.allocate(4).putInt(assertionResponse.getAuthenticatorData().getSignCount()).array(),
```

Comment:
This seems like it would only work in the case of U2F assertions.

Reply:
Assertion procedures across all types of attestations are identical.
Just make toBeSigned message with authData and hash of clientData, then verify received signature with saved credential public key.

Reply:
U2F assertions are signed over {app param, user presence, counter, challenge param} [https://fidoalliance.org/specs/fido-u2f-v1.2-ps-20170411/fido-u2f-raw-message-formats-v1.2-ps-20170411.html#authentication-response-message-success](url)

This implies that all types of attestations are not guaranteed to be the same, so I believe they should remain separated.

Reply:
@cpiper U2F raw messages are mapped to CTAP2 messages compatible to WebAutn Specs. So, if we get some responses from WebAuthn Javascript API with a U2F token, the responses are not same to legacy U2F format. See FIDO 2.0 CTAP specs.

Reply:
@Kieun, Sorry, I'm not following. Existing authenticators will always sign over that byte concatenation, so mapping it before sending to the server wouldn't change what bytes are signed.

Reply:
Signature is generated over same messages between legacy U2F and WebAuthn. I mean that the authData included in assertion response is already identical to the toBeSigned message for legacy U2F. So, the server does not have to construct the U2F message for it.

Reply:
The authenticator data returned from the FIDO 2.0 spec does not contain the challenge parameter. It should need to be reconstructed.

Reply:
Can you simply run the code and compare the original signBytes with authDataBytes? These are the same value.

Reply:
Right, I forgot that the client data hash is being used as the challenge parameter.